regnety/models/blocks.py

```diff
@@ -0,0 +1,219 @@
+"""Has building blocks of RegNetY models"""
+
+import tensorflow as tf
+
+
+# TODO:
+# 1. Stem
+# 2. Body
+#     2.1 Block
+#         2.1.1 SE
+# 3. Head
+# 4. RegNetY
+# Reference: https://tinyurl.com/dcc5c5p8
+
+class Stem(tf.keras.layers.Layer):
+    """Class to initiate stem architecture from the paper: 
+    `stride-two 3×3 conv with w0 = 32 output filters`
+    
+    Args:
+        None, stem is common to all models
+    """
+
+    def __init__(self):
+        super(Stem, self).__init__()
+        self.conv3x3 =  tf.keras.layers.Conv2D(32, (3,3), strides = 2)
+        self.bn = tf.keras.layers.BatchNormalization(
+            momentum = 0.9, epsilon = 0.00001
+        )
+        self.act = tf.keras.layers.ReLU()
+
+    def call(self, inputs):
+        x = self.conv3x3(inputs)
+        x = self.bn(x)
+        x = self.act(x)
+        return x
+
+class SE(tf.keras.layers.Layer):
+    """
+    Squeeze and Excite block. Takes se_ratio and in_filters as arguments. 
+
+    Args:
+        in_filters: Input filters. Output filters are equal to input filters 
+        se_ratio: Ratio for bottleneck filters
+    """
+
+    def __init__(self, 
+        in_filters: int,
+        se_ratio: float = 0.25
+    ):
+        super(SE, self).__init__()
+        
+        self.se_filters = int(in_filters * se_ratio)
+        self.out_filters = in_filters
+        self.ga_pool = tf.keras.layers.GlobalAveragePooling2D()
+        self.squeeze_dense = tf.keras.layers.Dense(self.se_filters, activation = 'relu')
+        self.excite_dense = tf.keras.layers.Dense(self.out_filters,  activation = 'sigmoid')
+        
+    def call(self, inputs):
+        # input shape: (h,w,out_filters)
+        x = self.ga_pool(inputs) #x: (out_filters)
+        x = self.squeeze_dense(x) #x: (se_filters)
+        x = self.excite_dense(x) #x: (out_filters)
+        x = tf.reshape(x, [-1,1,1,self.out_filters])
+        x = tf.math.multiply(x, inputs) #x: (h,w,out_filters)
+        return x
+        
+
+class YBlock(tf.keras.layers.Layer):
+    """
+    Y Block in RegNetY structure. 
+    IMPORTANT: Grouped convolutions are only supported by keras on GPU. 
```

Comment:
> Grouped convolutions are only supported by keras on GPU. 

Is this still true?

Reply:
I'm not aware of any new development regarding that. AFAIK, they're not supported on CPU. I'm sure that they work on GPU since I've tested that, but I'm not sure about TPU. I'll check that and update accordingly.

Reply:
I had provided a snippet. Did you try that?

Reply:
I think we agreed on using the Keras implementation itself. ([comment](https://github.com/AdityaKane2001/regnety/issues/4#issuecomment-870201961))
No I haven't filed an issue yet. I'll do that. Here's some additional info I found on this. Please see [this](https://github.com/tensorflow/tensorflow/blob/35c46d26cf420e7076abdcc3324f7a8b80f4710a/tensorflow/core/kernels/conv_ops.cc#L144-L226). Grouped convs are implemented in TF (albeit with bad performance) on CPU but they may not necessarily be a part of Keras, just as they were part of core TF before being added to Keras ([reference](https://github.com/tensorflow/tensorflow/pull/36773)).

Reply:
Oh cool.